File: src/java/org/voegtle/weatherstation/server/logic/WeatherDataRepair.java
package org.voegtle.weatherstation.server.logic;

import com.google.appengine.api.datastore.Key;
import org.voegtle.weatherstation.server.logic.data.RepairJob;
import org.voegtle.weatherstation.server.persistence.LocationProperties;
import org.voegtle.weatherstation.server.persistence.PersistenceManager;
import org.voegtle.weatherstation.server.persistence.SmoothedWeatherDataSet;

import java.util.*;
import java.util.logging.Logger;

public class WeatherDataRepair {
  private static final Logger log = Logger.getLogger(WeatherDataRepair.class.getName());

  private final PersistenceManager pm;
  private LocationProperties locationProperties;
  private List<SmoothedWeatherDataSet> datasets;

  public WeatherDataRepair(PersistenceManager pm, LocationProperties locationProperties) {
    this.pm = pm;
    this.locationProperties = locationProperties;
  }

  public List<SmoothedWeatherDataSet> repair(Date begin, Date end) {
    ArrayList<SmoothedWeatherDataSet> repaired = new ArrayList<>();
    WeatherDataFetcher weatherDataFetcher = new WeatherDataFetcher(pm, locationProperties);
    datasets = weatherDataFetcher.fetchSmoothedWeatherData(begin, end);

    repaired.addAll(removeDuplicates());

    datasets = weatherDataFetcher.fetchSmoothedWeatherData(begin, end);

    RepairJob next = getNextRepairJob();
    while (next.containsData()) {
      repair(next);
      repaired.addAll(next.getDefectDataSets());
      next = getNextRepairJob();
    }

    return repaired;
  }

  private Collection<SmoothedWeatherDataSet> removeDuplicates() {
    HashMap<Key, SmoothedWeatherDataSet> duplicate = new HashMap<>();
    SmoothedWeatherDataSet previousDataset = null;

    for (SmoothedWeatherDataSet dataset : datasets) {
      if (previousDataset != null) {
        if (dataset.getTimestamp().equals(previousDataset.getTimestamp())) {
          if (!dataset.isValid()) {
            duplicate.put(dataset.getKey(), dataset);
          } else {
            duplicate.put(previousDataset.getKey(), previousDataset);
          }
        }
      }
      previousDataset = dataset;
    }

    for (SmoothedWeatherDataSet dataset : duplicate.values()) {
      pm.removeDataset(dataset);
    }

    return duplicate.values();
  }

  private RepairJob getNextRepairJob() {
    RepairJob repairJob = new RepairJob();
    SmoothedWeatherDataSet previousDataset = null;
    for (SmoothedWeatherDataSet dataset : datasets) {
      if (!dataset.isValid()) {
        repairJob.setFirst(previousDataset);
        repairJob.addDefectDataSet(dataset);
      } else if (repairJob.containsData()) {
        repairJob.setLast(dataset);
        break;
      } else {
        previousDataset = dataset;
      }
    }

    repairJob.calculateStep();
    return repairJob;
  }

  private void repair(RepairJob repairJob) {
    if (repairJob.getFirst() != null) {
      SmoothedWeatherDataSet first = repairJob.getFirst();
      RepairJob.RepairStep step = repairJob.getStep();
      int index = 0;
      for (SmoothedWeatherDataSet ds : repairJob.getDefectDataSets()) {
        index++;
        log.info("repair " + index + " - " + ds.getTimestamp());
        log.info("insideTemperature: " + first.getInsideTemperature() + " " + step.insideTemperature);

        ds.setOutsideHumidity(getNewValue(first.getOutsideHumidity(), index, step.humidity));
        ds.setOutsideTemperature(getNewValue(first.getOutsideTemperature(), index, step.temperature));

        ds.setInsideHumidity(getNewValue(first.getInsideHumidity(), index, step.insideHumidity));
        ds.setInsideTemperature(getNewValue(first.getInsideTemperature(), index, step.insideTemperature));

        ds.setRainCounter(getNewValue(first.getRainCounter(), index, step.rain));
        ds.setKwh(getNewValue(first.getKwh(), index, step.kwh));

        setDefaults(ds);
        pm.updateDataset(ds);
      }
    }
  }

  private void setDefaults(SmoothedWeatherDataSet ds) {
    ds.setRaining(false);
    ds.setWindspeed((float) 0.0);
    ds.setWindspeedMax((float) 0.0);
    ds.setRepaired(true);
  }

  private Integer getNewValue(Integer startValue, int index, double step) {
    double value = startValue + index * step;
    return Math.round((float) value);
  }

  private Float getNewValue(Float startValue, int index, Double step) {
    log.info("startValue=" + startValue + ", step=" + step);
    if (step == null) {
      return null;
    }
    double value = startValue + index * step;
    log.info("value=" + value);
    return new Float(value);
  }

  private Double getNewValue(Double startValue, int index, Double step) {
    if (step == null) {
      return null;
    }
    return startValue + index * step;
  }
}

File: src/main/java/org/voegtle/weatherstation/server/logic/WeatherDataRepair.kt
package org.voegtle.weatherstation.server.logic

import org.voegtle.weatherstation.server.logic.data.RepairJob
import org.voegtle.weatherstation.server.persistence.PersistenceManager
import org.voegtle.weatherstation.server.persistence.entities.LocationProperties
import org.voegtle.weatherstation.server.persistence.entities.SmoothedWeatherDataSet2
import java.util.*
import java.util.logging.Logger

class WeatherDataRepair(private val pm: PersistenceManager, private val locationProperties: LocationProperties) {
  private var datasets: MutableList<SmoothedWeatherDataSet2> = ArrayList<SmoothedWeatherDataSet2>()

  private fun fetchNextRepairJob(): RepairJob {
      val repairJob = RepairJob()
      var previousDataset: SmoothedWeatherDataSet2? = null
      val iter = datasets.iterator()
      while (iter.hasNext()) {
        val dataset = iter.next()
        if (!dataset.isValid) {
          repairJob.first = previousDataset
          repairJob.addDefectDataSet(dataset)
          iter.remove()
        } else if (repairJob.containsData()) {
          repairJob.last = dataset
          break
        } else {
          previousDataset = dataset
        }
      }

      repairJob.calculateStep()
      return repairJob
    }

  fun repair(begin: Date, end: Date?): List<SmoothedWeatherDataSet2> {
    val repaired = ArrayList<SmoothedWeatherDataSet2>()
    val weatherDataFetcher = WeatherDataFetcher(pm, locationProperties)
    datasets = weatherDataFetcher.fetchSmoothedWeatherData(begin, end)

    repaired.addAll(removeDuplicates())

    datasets = weatherDataFetcher.fetchSmoothedWeatherData(begin, end)

    var next = fetchNextRepairJob()
    while (next.containsData()) {
      repair(next)
      repaired.addAll(next.defectDataSets)
      next = fetchNextRepairJob()
    }

    return repaired
  }

  private fun removeDuplicates(): Collection<SmoothedWeatherDataSet2> {
    val duplicate = HashMap<Key, SmoothedWeatherDataSet2>()
    var previousDataset: SmoothedWeatherDataSet2? = null

    for (dataset in datasets) {
      if (previousDataset != null) {
        if (dataset.timestamp == previousDataset.timestamp) {
          if (!dataset.isValid) {
            duplicate.put(dataset.id!!, dataset)
          } else {
            duplicate.put(previousDataset.id!!, previousDataset)
          }
        }
      }
      previousDataset = dataset
    }

    for (dataset in duplicate.values) {
      pm.removeDataset(dataset)
    }

    return duplicate.values
  }

  private fun repair(repairJob: RepairJob) {
    repairJob.first?.let {
      val step = repairJob.step
      var index = 0
      for (ds in repairJob.defectDataSets) {
        index++
        log.info("repair " + index + " - " + ds.timestamp)
        log.info("insideTemperature: " + it.insideTemperature + " " + step.insideTemperature)

        ds.outsideHumidity = getNewValue(it.outsideHumidity, index, step.humidity)
        ds.outsideTemperature = getNewValue(it.outsideTemperature, index, step.temperature)

        ds.insideHumidity = getNewValue(it.insideHumidity, index, step.insideHumidity)
        ds.insideTemperature = getNewValue(it.insideTemperature, index, step.insideTemperature)

        it.rainCounter?.let { rainCounter -> ds.rainCounter = getNewValue(rainCounter, index, step.rain) }

        it.kwh?.let {
          ds.kwh = getNewValue(it, index, step.kwh)
        }

        setDefaults(ds)
        pm.updateDataset(ds)
      }
    }
  }

  private fun setDefaults(ds: SmoothedWeatherDataSet2) {
    ds.isRaining = false
    ds.windspeed = 0.0.toFloat()
    ds.windspeedMax = 0.0.toFloat()
    ds.repaired = true
  }

  private fun getNewValue(startValue: Int, index: Int, step: Double): Int {
    val value = startValue + index * step
    return Math.round(value).toInt()
  }

  private fun getNewValue(startValue: Float?, index: Int, step: Double?): Float? {
    if (step == null || startValue == null) {
      return null
    }
    val value = startValue + index * step
    return value.toFloat()
  }

  private fun getNewValue(startValue: Double, index: Int, step: Double?): Double? {
    return if (step == null) {
      null
    } else startValue + index * step
  }

  companion object {
    private val log = Logger.getLogger(WeatherDataRepair::class.java.name)
  }
}
